flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSource.java

```diff
@@ -0,0 +1,209 @@
+                t -> serializers.add(castSerializer(t.f0.getEnumeratorCheckpointSerializer())));
+        return new HybridSourceEnumStateSerializer(serializers);
+    }
+
+    private static <T> SimpleVersionedSerializer<T> castSerializer(
+            SimpleVersionedSerializer<? extends T> s) {
+        @SuppressWarnings("rawtypes")
+        SimpleVersionedSerializer s1 = s;
+        return s1;
+    }
+
+    /** Serializes splits by delegating to the source-indexed split serializer. */
+    public static class SplitSerializerWrapper<SplitT extends SourceSplit>
+            implements SimpleVersionedSerializer<HybridSourceSplit> {
+
+        final List<SimpleVersionedSerializer<SourceSplit>> serializers;
+
+        public SplitSerializerWrapper(List<SimpleVersionedSerializer<SourceSplit>> serializers) {
+            this.serializers = serializers;
+        }
+
+        @Override
+        public int getVersion() {
+            return 0;
+        }
+
+        @Override
+        public byte[] serialize(HybridSourceSplit split) throws IOException {
+            try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                    DataOutputStream out = new DataOutputStream(baos)) {
+                out.writeInt(split.sourceIndex());
+                out.writeInt(serializerOf(split.sourceIndex()).getVersion());
+                byte[] serializedSplit =
+                        serializerOf(split.sourceIndex()).serialize(split.getWrappedSplit());
+                out.writeInt(serializedSplit.length);
+                out.write(serializedSplit);
+                return baos.toByteArray();
+            }
+        }
+
+        @Override
+        public HybridSourceSplit deserialize(int version, byte[] serialized) throws IOException {
+            try (ByteArrayInputStream bais = new ByteArrayInputStream(serialized);
+                    DataInputStream in = new DataInputStream(bais)) {
+                int sourceIndex = in.readInt();
+                int nestedVersion = in.readInt();
+                int length = in.readInt();
+                byte[] splitBytes = new byte[length];
+                in.readFully(splitBytes);
+                SourceSplit split =
+                        serializerOf(sourceIndex).deserialize(nestedVersion, splitBytes);
+                return new HybridSourceSplit(sourceIndex, split);
+            }
+        }
+
+        private SimpleVersionedSerializer<SourceSplit> serializerOf(int sourceIndex) {
+            Preconditions.checkArgument(sourceIndex < serializers.size());
+            return serializers.get(sourceIndex);
+        }
+    }
+
+    /**
+     * Converts checkpoint between sources to transfer end position to next source's start position.
+     * Only required for dynamic position transfer at time of switching, otherwise source can be
+     * preconfigured with a start position during job submission.
+     */
+    public interface CheckpointConverter<InCheckpointT, OutCheckpointT>
```

Comment:
IMO, `CheckpointConverter` interface doesn't need to be defined, which only extends the `Function` interface.